Classes in method arguments are not put into extension jar

## Expected behavior
A class that is used in as a method parameter should also be put into the extension jar.
The causes the extension to throw a ClassNotFoundException.

## Actual behavior
The class is missing.

### Reproducer code

```
public class MyExtension implements ExtensionMain {

    @Override
    public void extensionStart(
           final @NotNull ExtensionStartInput extensionStartInput, 
           final @NotNull ExtensionStartOutput extensionStartOutput) {
                   doNothing(new B());
    }

    private void doNothing(final @NotNull A a) {

    }
}

public class B extends A {
}

public class A {
}
````

throws:

```
Caused by: java.lang.ClassNotFoundException: A
	at java.base/java.net.URLClassLoader.findClass(Unknown Source)
	at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
	at com.hivemq.extensions.classloader.IsolatedPluginClassloader.loadClass(IsolatedPluginClassloader.java:123)
	at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
```

## Details
- Affected HiveMQ Testcontainer version(s): 1.0.0

--- a/core/src/main/java/com/hivemq/testcontainer/core/HiveMQTestContainerCore.java
+++ b/core/src/main/java/com/hivemq/testcontainer/core/HiveMQTestContainerCore.java
@@ -35,7 +35,7 @@ import java.io.File;
 import java.io.IOException;
 import java.nio.charset.Charset;
 import java.time.Duration;
-import java.util.Set;
+import java.util.Collection;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.TimeUnit;
@@ -239,9 +239,8 @@ public class HiveMQTestContainerCore<SELF extends HiveMQTestContainerCore<SELF>>
             final @NotNull JavaArchive javaArchive) throws NotFoundException {
 
         if (clazz != null) {
-            final Set<String> subClassNames =
-                    ClassPool.getDefault().get(clazz.getName()).getClassFile().getConstPool().getClassNames();
-            for (final String subClassName : subClassNames) {
+            final Collection<String> refClasses = ClassPool.getDefault().get(clazz.getName()).getRefClasses();
+            for (final String subClassName : refClasses) {
                 final String className = subClassName.replaceAll("/", ".");
                 logger.debug("Packaging subclass {} into extension {}.", className, extensionId);
                 javaArchive.addClass(className);

--- a/core/src/main/java/com/hivemq/testcontainer/core/HiveMQTestContainerCore.java
+++ b/core/src/main/java/com/hivemq/testcontainer/core/HiveMQTestContainerCore.java
@@ -240,7 +240,7 @@ public class HiveMQTestContainerCore<SELF extends HiveMQTestContainerCore<SELF>>
 
         if (clazz != null) {
             final Collection<String> refClasses = ClassPool.getDefault().get(clazz.getName()).getRefClasses();
-            for (final String subClassName : refClasses) {
+            for (final String refClass : refClasses) {
                 final String className = subClassName.replaceAll("/", ".");
                 logger.debug("Packaging subclass {} into extension {}.", className, extensionId);
                 javaArchive.addClass(className);

--- a/core/src/main/java/com/hivemq/testcontainer/core/HiveMQTestContainerCore.java
+++ b/core/src/main/java/com/hivemq/testcontainer/core/HiveMQTestContainerCore.java
@@ -241,7 +241,7 @@ public class HiveMQTestContainerCore<SELF extends HiveMQTestContainerCore<SELF>>
         if (clazz != null) {
             final Collection<String> refClasses = ClassPool.getDefault().get(clazz.getName()).getRefClasses();
             for (final String refClass : refClasses) {
-                final String className = subClassName.replaceAll("/", ".");
+                final String className = refClass.replaceAll("/", ".");
                 logger.debug("Packaging subclass {} into extension {}.", className, extensionId);
                 javaArchive.addClass(className);
             }

--- a/core/src/test/java/com/hivemq/testcontainer/util/A.java
+++ b/core/src/test/java/com/hivemq/testcontainer/util/A.java
@@ -15,5 +15,8 @@
  */
 package com.hivemq.testcontainer.util;
 
+/**
+ * @author Yannick Weber
+ */
 public class A {
 }

--- a/core/src/test/java/com/hivemq/testcontainer/util/A.java
+++ b/core/src/test/java/com/hivemq/testcontainer/util/A.java
@@ -1,0 +1,19 @@
+/*
+ * Copyright 2020 HiveMQ and the HiveMQ Community
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hivemq.testcontainer.util;
+
+public class A {
+}

--- a/core/src/test/java/com/hivemq/testcontainer/util/B.java
+++ b/core/src/test/java/com/hivemq/testcontainer/util/B.java
@@ -15,5 +15,8 @@
  */
 package com.hivemq.testcontainer.util;
 
+/**
+ * @author Yannick Weber
+ */
 public class B extends A {
 }

--- a/core/src/test/java/com/hivemq/testcontainer/util/B.java
+++ b/core/src/test/java/com/hivemq/testcontainer/util/B.java
@@ -1,0 +1,19 @@
+/*
+ * Copyright 2020 HiveMQ and the HiveMQ Community
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hivemq.testcontainer.util;
+
+public class B extends A {
+}

--- a/core/src/test/java/com/hivemq/testcontainer/util/MyExtensionWithArgumentClasses.java
+++ b/core/src/test/java/com/hivemq/testcontainer/util/MyExtensionWithArgumentClasses.java
@@ -1,0 +1,61 @@
+/*
+ * Copyright 2020 HiveMQ and the HiveMQ Community
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hivemq.testcontainer.util;
+
+import com.hivemq.extension.sdk.api.ExtensionMain;
+import com.hivemq.extension.sdk.api.annotations.NotNull;
+import com.hivemq.extension.sdk.api.interceptor.publish.PublishInboundInterceptor;
+import com.hivemq.extension.sdk.api.parameter.ExtensionStartInput;
+import com.hivemq.extension.sdk.api.parameter.ExtensionStartOutput;
+import com.hivemq.extension.sdk.api.parameter.ExtensionStopInput;
+import com.hivemq.extension.sdk.api.parameter.ExtensionStopOutput;
+import com.hivemq.extension.sdk.api.services.Services;
+import com.hivemq.extension.sdk.api.services.intializer.ClientInitializer;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * @author Yannick Weber
+ */
+@SuppressWarnings("CodeBlock2Expr")
+public class MyExtensionWithArgumentClasses implements ExtensionMain {
+
+    @Override
+    public void extensionStart(final @NotNull ExtensionStartInput extensionStartInput, final @NotNull ExtensionStartOutput extensionStartOutput) {
+
+        final PublishInboundInterceptor publishInboundInterceptor = (publishInboundInput, publishInboundOutput) -> {
+            publishInboundOutput.getPublishPacket().setPayload(ByteBuffer.wrap("modified".getBytes(StandardCharsets.UTF_8)));
+        };
+
+        final ClientInitializer clientInitializer = (initializerInput, clientContext) -> {
+            clientContext.addPublishInboundInterceptor(publishInboundInterceptor);
+        };
+
+        Services.initializerRegistry().setClientInitializer(clientInitializer);
+
+        doNothing(new B());
+    }
+
+    @Override
+    public void extensionStop(final @NotNull ExtensionStopInput extensionStopInput, final @NotNull ExtensionStopOutput extensionStopOutput) {
+
+    }
+
+    public void doNothing(final @NotNull A a) {
+
+    }
+}

--- a/junit4/src/test/java/com/hivemq/testcontainer/junit4/ContainerWithExtensionWithArgumentClassesIT.java
+++ b/junit4/src/test/java/com/hivemq/testcontainer/junit4/ContainerWithExtensionWithArgumentClassesIT.java
@@ -1,0 +1,48 @@
+/*
+ * Copyright 2020 HiveMQ and the HiveMQ Community
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hivemq.testcontainer.junit4;
+
+import com.hivemq.extension.sdk.api.annotations.NotNull;
+import com.hivemq.testcontainer.core.HiveMQExtension;
+import com.hivemq.testcontainer.util.MyExtensionWithArgumentClasses;
+import com.hivemq.testcontainer.util.TestPublishModifiedUtil;
+import org.junit.Rule;
+import org.junit.Test;
+
+import java.util.concurrent.ExecutionException;
+
+/**
+ * @author Yannick Weber
+ */
+public class ContainerWithExtensionWithArgumentClassesIT {
+
+    @Rule
+    public final @NotNull HiveMQTestContainerRule rule =
+            new HiveMQTestContainerRule()
+                    .withExtension(HiveMQExtension.builder()
+                            .id("extension-1")
+                            .name("my-extension")
+                            .version("1.0")
+                            .mainClass(MyExtensionWithArgumentClasses.class).build());
+
+
+    @Test(timeout = 500_000)
+    public void test_extension_with_argument_classes() throws ExecutionException, InterruptedException {
+        TestPublishModifiedUtil.testPublishModified(rule.getMqttPort());
+    }
+
+
+}

--- a/junit5/src/test/java/com/hivemq/testcontainer/junit5/ContainerWithExtensionWithArgumentClassesIT.java
+++ b/junit5/src/test/java/com/hivemq/testcontainer/junit5/ContainerWithExtensionWithArgumentClassesIT.java
@@ -1,0 +1,50 @@
+/*
+ * Copyright 2020 HiveMQ and the HiveMQ Community
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hivemq.testcontainer.junit5;
+
+import com.hivemq.extension.sdk.api.annotations.NotNull;
+import com.hivemq.testcontainer.core.HiveMQExtension;
+import com.hivemq.testcontainer.util.MyExtensionWithArgumentClasses;
+import com.hivemq.testcontainer.util.TestPublishModifiedUtil;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.Timeout;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * @author Yannick Weber
+ */
+public class ContainerWithExtensionWithArgumentClassesIT {
+
+    @RegisterExtension
+    public final @NotNull HiveMQTestContainerExtension extension =
+            new HiveMQTestContainerExtension()
+                    .withExtension(HiveMQExtension.builder()
+                            .id("extension-1")
+                            .name("my-extension")
+                            .version("1.0")
+                            .mainClass(MyExtensionWithArgumentClasses.class).build());
+
+    @Test()
+    @Timeout(value = 5, unit = TimeUnit.MINUTES)
+    void test_extension_with_argument_class() throws ExecutionException, InterruptedException {
+        TestPublishModifiedUtil.testPublishModified(extension.getMqttPort());
+    }
+
+
+}

